liquibase/liquibase
# Handle the fact that FK names are not always unique.

## Description
Update ForeignKeyComparator logic to work like PrimaryKeyComparitor which already assumes primary key names are only unique within the table name.

Fixes #2227 

- - -
## Dev Handoff Notes (Internal Use)
#### Links
* Fixed Issue: [https://github.com/liquibase/liquibase/issues/2227](https://github.com/liquibase/liquibase/issues/2227) 
* Test Results: [https://github.com/liquibase/liquibase/pull/2228/checks](https://github.com/liquibase/liquibase/pull/2228/checks)

#### Testing
* Steps to Reproduce: [https://github.com/liquibase/liquibase/issues/2227](https://github.com/liquibase/liquibase/issues/2227) 
* Guidance:
  * Impact: Logic used for uniquely identifying foreign keys in snapshot/diff/diffChangeLog logic

#### Dev Verification
With the change, I get

```java
Missing Foreign Key(s): 
     fk_test_fk(table2[id] -> table1[id])
```

## Test Requirements (Liquibase Internal QA)
You need two Postgres instances to test this fix. Postgres version does not matter; I reproed using a PG 12 and a PG 9. 

**Setup:**

On one PG instance, execute the following SQL:

```java
create table table1 (id int primary key);
create table table2 (id int primary key);
alter table table1 add constraint fk_test_fk FOREIGN KEY (id) references table1(id);
alter table table2 add constraint fk_test_fk FOREIGN KEY (id) references table1(id);
```

On the second PG instance, execute the following SQL:

```java
create table table1 (id int primary key);
create table table2 (id int primary key);
alter table table1 add constraint fk_test_fk FOREIGN KEY (id) references table1(id);
```

The difference is the missing foreign key on table2 on the second instance.

**Manual Tests**

_Verify liquibase diff shows a unexpected foreign key fk_test_fk from table2(id) to table1(id)._

* In liquibase.properties, specify url as the first database and referenceUrl as the second database.
* liquibase diff --schemas public

_Verify liquibase diffChangeLog has an ALTER TABLE to drop a foreign key fk_test_fk from table2(id) to table1(id)._

* In liquibase.properties, specify url as the first database and referenceUrl as the second database.
* liquibase diff-changelog --changelog-file gcl_dropFk.postgres.sql --schemas public
  * The SQL output to the changelog is:
  * `ALTER  TABLE  table2  DROP  FOREIGN  KEY  fk_test_fk;`

_Verify reversing the order of the diff shows an missing foreign key fk_test_fk from table2(id) to table1(id)._

* In liquibase.properties, specify url as the second database and referenceUrl as the first database.
* liquibase diff --schemas public

_Verify reversing the order of databases during diff-changelog generates a  ALTER TABLE to add a foreign key fk_test_fk from table2(id) to table1(id)._

* In liquibase.properties, specify url as the second database and referenceUrl as the first database.
* liquibase diff-changelog --changelog-file gcl_addFk.postgres.sql --schemas public
  * The SQL output to the changelog is:
  * `alter table table2 add constraint fk_test_fk FOREIGN KEY (id) references table1(id);`

_Verify liquibase update using changelog gcl_dropFK.postgres.sql correctly drops the foreign key from the first database._

(Note: I added this test case because I have a suspicion that if the SQL to drop the FK will throw an error on update because there are two “fk_test_fk” foreign keys on table1. If there is no error during update itself, verify that the correct/expected foreign key is dropped.)

* In liquibase.properties, specify url as the first database. (In this test referenceUrl is not used).
* liquibase update --changelog-file gcl_dropFk.postgres.sql
  * Update is successful.
  * Validation directly on the database confirms the foreign key on table1 from table2 is dropped.

**Automated Tests**

* None. There are sufficient integration/dev tests for testing foreign keys.



┆Issue is synchronized with this [Jira Bug](https://datical.atlassian.net/browse/LB-2176) by [Unito](https://www.unito.io)
